File: 322.零钱兑换.py
#
# @lc app=leetcode.cn id=322 lang=python
#
# [322] 零钱兑换
#

# @lc code=start
class Solution(object):
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        coins = sorted(coins)

        if amount==0:return 0
        if amount < coins[0]: return -1
        res = [0]
        res.extend([ -1 for i in range(coins[0]-1)])
        # res.append(1)
        for i in range(coins[0], amount+1):
            tmp_res = 10e5
            for c in coins:
                if i-c<0: break
                if res[i-c]<0: break
                tmp_res = tmp_res if res[i-c]+1 > tmp_res else res[i-c]+1
            if tmp_res==10e5:tmp_res=-1
            res.append(tmp_res)
            # if tmp_res>0:print(res)
        # print(res)
        return res[-1]
            
        
    
# @lc code=end

File: 322.零钱兑换.js
/*
 * @lc app=leetcode.cn id=322 lang=javascript
 *
 * [322] 零钱兑换
 */

// @lc code=start
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */
var coinChange = function(coins, amount) {

    let a=new Array(amount+1).fill(-1);
    a[0]=0;

    for(let i=1;i<=amount;i++){
        for(let j=0;j<coins.length;j++){
            // console.log(i,coins[j])
            
            if(i-coins[j]>=0 && a[i-coins[j]]!=-1){
                if(a[i]===-1){
                    a[i]=a[i-coins[j]]+1;
                }else{
                    a[i]=Math.min(a[i-coins[j]]+1,a[i]);
                }
               
            }
        }
 
    }

    return a[amount];
};
// @lc code=end

